pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -368,7 +368,18 @@ async def _prepare_request(
 
         # Check usage
         if ctx.deps.usage_limits:  # pragma: no branch
-            ctx.deps.usage_limits.check_before_request(ctx.state.usage)
+            if ctx.deps.usage_limits.count_tokens_before_request:
+                model_request_parameters = await _prepare_request_parameters(ctx)
+                message_history = await _process_message_history(
+                    ctx.state.message_history, ctx.deps.history_processors, build_run_context(ctx)
+                )
+                token_count = await ctx.deps.model.count_tokens(
+                    message_history, ctx.deps.model_settings, model_request_parameters
+                )
+                ctx.state.usage.incr(token_count.to_usage())
```

Comment:
I don't think we should add this to the actual `ctx.state.usage`, as it would end up double counting the estimated usage with the real usage added later. Instead, we can copy `ctx.state.usage` (possibly with just a plain `dataclasses.replace(ctx.state_usage)`, add it there, and then check against that